### ISSUE TITLE
Allocation via new/delete when using custom allocator

### ISSUE THREAD

**Issue by mojofunk** [fmtlib/fmt#728]
I've been using libfmt for a few years now as part of a logging/instrumentation library I developed and I have generally been very happy with it.

I was primarily developing and using my library on linux and using gcc as the compiler. I recently ported it to Windows and started using the compiler that comes with Visual Studio 2017.

It was fairly straight forward and not too many changes were necessary to my library. There was one change in the implementation of the C++ libraries that come with Visual Studio 2017 that causes an issue for me and has made me realize that an assumption I had made in regards to the libfmt API was not correct.

I am using a custom allocator and string type as described in http://fmtlib.net/latest/api.html#custom-allocators and I assumed that this would mean that the custom allocator would be used for any
allocation that was required during formatting.

It seems like that is not the case, at least in versions 3.0 and 4.1 that I have tested. I've attached a basic unit test that demonstrates the issue if compiled with Visual Studio 2017(and potentially other
compilers). It overrides the global c++ allocators (via new/delete) and then disables allocation during the formatting.

The unit test case will pass if compiled with gcc (version 7.2) and I'm guessing this is due to the implementation of std::vector and its different allocation behaviour. Hopefully it isn't due to bugs in my
unit test.

So I guess my question is whether libfmt is intended to support my use-case and guarantee that all allocation is performed via the custom allocator and not use a std::allocator/new/malloc/etc?

[fmt_alloc_test.cpp.txt](https://github.com/fmtlib/fmt/files/1976358/fmt_alloc_test.cpp.txt)

**Comment by mojofunk** [comment]
Here is the callstack which shows where the allocation is occurring during the test when built using the compiler/C++ library that comes with Visual Studio 2017:

fmt_alloc_test.exe!std::_Default_allocator_traits<std::allocator<std::_Container_proxy> >::construct<std::_Container_proxy,std::_Container_proxy>(std::allocator<std::_Container_proxy> & __formal, std::_Container_proxy * const _Ptr, std::_Container_proxy && <_Args_0>) Line 918
fmt_alloc_test.exe!std::_Vector_alloc<std::_Vec_base_types<std::pair<fmt::BasicStringRef<char>,fmt::internal::Arg>,std::allocator<std::pair<fmt::BasicStringRef<char>,fmt::internal::Arg> > > >::_Alloc_proxy() Line 552
fmt_alloc_test.exe!std::_Vector_alloc<std::_Vec_base_types<std::pair<fmt::BasicStringRef<char>,fmt::internal::Arg>,std::allocator<std::pair<fmt::BasicStringRef<char>,fmt::internal::Arg> > > >::_Vector_alloc<std::_Vec_base_types<std::pair<fmt::BasicStringRef<char>,fmt::internal::Arg>,std::allocator<std::pair<fmt::BasicStringRef<char>,fmt::internal::Arg> > > >() Line 496
fmt_alloc_test.exe!std::vector<std::pair<fmt::BasicStringRef<char>,fmt::internal::Arg>,std::allocator<std::pair<fmt::BasicStringRef<char>,fmt::internal::Arg> > >::vector<std::pair<fmt::BasicStringRef<char>,fmt::internal::Arg>,std::allocator<std::pair<fmt::BasicStringRef<char>,fmt::internal::Arg> > >() Line 689
fmt_alloc_test.exe!fmt::internal::ArgMap<char>::ArgMap<char>()
fmt_alloc_test.exe!fmt::BasicFormatter<char,fmt::ArgFormatter<char> >::BasicFormatter<char,fmt::ArgFormatter<char> >(const fmt::ArgList & args, fmt::BasicWriter<char> & w) Line 2329
fmt_alloc_test.exe!fmt::BasicWriter<char>::write(fmt::BasicCStringRef<char> format, fmt::ArgList args) Line 2781
fmt_alloc_test.exe!test_format(TestStringAllocator<char> alloc, fmt::BasicCStringRef<char> format_str, fmt::ArgList args) Line 187
fmt_alloc_test.exe!test_format<unsigned int,char const *>(TestStringAllocator<char> arg0, fmt::BasicCStringRef<char> arg1, const unsigned int & <args_0>, const char * const & <args_1>) Line 190
fmt_alloc_test.exe!FmtAllocTest_FmtAllocTestBasic_Test::TestBody() Line 201
gtestd.dll!testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test,void>(testing::Test * object, void(testing::Test::*)() method, const char * location) Line 2387
gtestd.dll!testing::internal::HandleExceptionsInMethodIfSupported<testing::Test,void>(testing::Test * object, void(testing::Test::*)() method, const char * location) Line 2438
gtestd.dll!testing::Test::Run() Line 2474

**Comment by vitaut** [comment]
> So I guess my question is whether libfmt is intended to support my use-case and guarantee that all allocation is performed via the custom allocator and not use a std::allocator/new/malloc/etc?

[Custom allocator support](http://fmtlib.net/latest/api.html#custom-allocators) allows you to pass an allocator for the output buffer. However, it doesn't guarantee that all allocations will be done through it. In particular, if you are using named arguments, the vector holding the mapping from names to argument values will use the default allocator. (If you are not using named arguments no allocation will be done.) Also floating-point formatting falls back to `s(n)printf` which may allocate memory under certain conditions.

**Comment by mojofunk** [comment]
> Custom allocator support allows you to pass an allocator for the output buffer. However, it doesn't guarantee that all allocations will be done through it.

OK, That is a bit of a shame, but understandable. Some additional documentation on that page to that effect might be useful and appreciated.

> In particular, if you are using named arguments, the vector holding the mapping from names to argument values will use the default allocator. (If you are not using named arguments no allocation will be done.)

I don't believe I was using named arguments in the unit test that I attached or floating point formatting. As I said, I only discovered this issue when using Visual Studio 2017 compiler. The std::vector ctor is allocating using new via the BasicFormatter ctor with that compiler, at least in a debug config.

I notice that the implementation of BasicFormatter/ArgMap or now basic_formatter_context/arg_map is no longer a std::vector but a dynamically allocated array in the master branch. I'm not that familiar with the code, so I find it all a bit hard to follow but it looks like in master that the arg_map.map_ member is only initialized via arg_map<>::init if indexed arguments or named arguments are used.

So for instance at least in master (ca31ca13f) something like:

format("{0}, {1}, {2}", 'a', 'b', 'c');

would allocate via new to create the arg_map.map_ member, but:

format("{}, {}, {}", 'a', 'b', 'c');

would not.

I have not used named arguments before but it looks like they are used something like:

format("{arg2} : {arg1}", fmt::arg("arg1", 42), fmt::arg("arg2", "The answer to life, the universe and everything"));

And that would also allocate the arg_map.map_ member via new?

I don't use positional (indexed) or named arguments, so if you could confirm the above that would be appreciated.

But from what you have said it sounds like there are really no guarantees with regards to allocation and the implementation may change. This is acceptable to me and the fmt library is still fantastic but some additional documentation to make it clear would be appreciated.

**Comment by vitaut** [comment]
I clarified the use of allocators in https://github.com/fmtlib/fmt/commit/f0ae725721d766842dde88ed70ac8bea6085b36f.

Also, it appears that MSVC may allocate memory for an empty `std::vector` which explains why you see allocations even when named arguments are not used. In the `master` branch, `std::vector` is replaced with an array manually allocated with `new` so the problem on MSVC should go away.

In your examples only

```c++
format("{arg2} : {arg1}", fmt::arg("arg1", 42),
       fmt::arg("arg2", "The answer to life, the universe and everything"));
```
should allocate on `master`.

Thanks for reporting the issue.
